HIP/hip-904-frictionless-airdrop.md

```diff
@@ -266,194 +454,334 @@ Note: the only change to this API is in the specification, which now allows for
+     * This list MUST contain between 1 and 10 transfers, inclusive.
+     * <p>
+     * <blockquote>Note that each transfer of fungible/common tokens requires both a debit and
+     * a credit, so each _fungible_ token transfer MUST have _balanced_ entries in the
+     * TokenTransferList for that transfer.</blockquote>
+     */
     repeated TokenTransferList token_transfers = 1;
 }
 ```
 
-TokenAirdrop transaction will support the transfers of value in a similar manner to CryptoTransfer with the difference of only supporting tokens (no HBAR transfer support) and supporting the resulting pending transfer state where a CryptoTransfer would have failed due to the lack of association options.
+TokenAirdrop transaction will support the transfers of value in a similar manner to CryptoTransfer
+with the difference of only supporting tokens (no HBAR transfer support) and supporting the
+resulting pending transfer state where a CryptoTransfer would have failed due to the lack of
+association options.
 
 However, it is important to have a separate HAPI transaction type for a few reasons
 
-- the ability to call out the potential separate behaviours based on maxAutomaticAssociation, whereby transfer can end up in a pending transfer state or a consumption of an association slot.
+- the ability to call out the potential separate behaviours based on maxAutomaticAssociation,
+  whereby transfer can end up in a pending transfer state or a consumption of an association slot.
 - cleaner conceptual understanding
-- naming allows the grouping of functionality with Cancel and Claim which complete the pending state a TokenAirdrop may creates
+- naming allows the grouping of functionality with Cancel and Claim which complete the pending
+  state a TokenAirdrop may creates
 - explicit specification in record stream.
 - improved developer experience
 
-Similar restrictions around transfer list and aggregation across sender and receiver from CryptoTransfer will be applied to TokenAirdop e.g. max 10 transfers per transaction
+Similar restrictions around transfer list and aggregation across sender and receiver from
+CryptoTransfer will be applied to TokenAirdop e.g. max 10 transfers per transaction
 
-At the time of this HIP `TokenAirdropTransaction` will not be supported by the Scheduled Transaction Service. A future HIP may explore this support.
+At the time of this HIP `TokenAirdropTransaction` will not be supported by the Scheduled
+Transaction Service. A future HIP may explore this support.
 
-#### Cancel
+#### Pending Airdrop
 
 ```protobuf
-message TokenPendingAirdrop {
+/**
+ * A unique, composite, identifier for a pending airdrop.
+ *
+ * Each pending airdrop SHALL be uniquely identified by a PendingAirdropId.
+ * A PendingAirdropId SHALL be recorded when created and MUST be provided in any transaction
+ * that would modify that pending airdrop (such as a `claimAirdrop` or `cancelAirdrop`).
+ */
+message PendingAirdropId {
```

Comment:
Usually we only highlight the input and output how protobufs, however, as you've pointed out state protobuf needs to be added now.
In this case we should call out which components are for transaction input, storage and record output

As written I think you're suggesting
Transaction input -  PendingAirdropId 
Storage -  PendingAirdropId -> ? (maybe  PendingAirdrop)
Transaction Record output -  PendingAirdropRecord

However, `PendingAirdropId` probably doesn't need the reference to the indirect value prop serial_id. This makes it tough reuse in state.
Also not sure of the need of `PendingAirdrop` when it only has `amount`, why not just have amount in `PendingAirdropRecord` instead? 

What if we consider the following
- `TokenAirdropTransaction`: Input - `TokenTransferList`, Output: `repeated PendingAirdrop`
- `TokenCancelAirdrop`: Input - `repeated PendingAirdropId`
- `TokenClaimAirdrop`: Input - `repeated PendingAirdrop` (since a subset of nfts needs to be supported)
- State: `PendingAirdropId` -> `TokenValue`


```protobuf

message PendingAirdropId {
    /**
    * Account that issue the airdrop trasnaction
    */
    AccountID sender_id = 1;

	/**
    * Account that was the intended receiver of the airdrop
    */
    AccountID receiver_id = 2;

	/**
    * Fungible or Non Fungible token ID.
	* Amount or serials are captured in the TokenPendingAirdrop record
    */
    TokenID token_id = 3;	
}

/**
* Storage object to capture the amount of value held in a balance or pending transfer state
*/
message TokenValue {
	/**
    * The token under consideration
    */
	oneof token_value {
		int64 amount = 1;
		repeated int64 serials_ids = 2;
	}	
}


message PendingAirdrop {
    /**
     * A unique, composite, identifier for a pending airdrop.
     * This field is REQUIRED.
     */
    PendingAirdropId pending_airdrop_id = 1;

    /**
     * Pending NF airdrop serials.<br/>
     * This field is null for fungible tokens but required for NFTs
     */
    TokenValue token_value = 2;
}

```



```protobuf
/**
 * A token cancel will allow an airdrop sender to revert clear a pending transfer from state.
 * Afer this no pending transfer will exist for a potential receiver to claim.  
 * This transfer will be represented in the transaction body and will not need restating in the record file.
 * All cancellations must succeed for the transaction to succeed.
 */
message TokenCancelAirdropTransactionBody {
    /**
    * A list of up to 10 pending airdrops that may be cancelled. If there are
    * duplicate pending_airdrops, then the transaction will fail.
    */
    repeated PendingAirdropId pending_airdrops = 1;
}
```

#### Claim

```protobuf
/**
 * A token cancel will clear a pending transfer from state. 
 * This transfer will be represented in the transaction body and will not need restating in the record file.
 * All claims must succeed for the transaction to succeed.
 */
message TokenClaimAirdropTransactionBody {
    /**
    * The pending airdrop to be accepted. The transaction must be signed by
    * the `receiverAccountID`'s key as contained in the pending airdrop.
    * A list of up to 10 pending airdrops that may be claimed. If there are
    * duplicate pending_airdrops, then the transaction will fail.
    */
    repeated PendingAirdrop pending_airdrops = 1;
}
```

#### TransactionRecord

```protobuf
message TransactionRecord {

    ...

    /**
    * All pending token transfers created while handling this transaction. This
    * is the result of airdrops to account with no open association slots
    */
    repeated PendingAirdrop new_pending_airdrops = 22;
}
```

Reply:
Regarding usage
PendingAirdropId is a key
PendingAirdrop is the value in the map for that key
PendingAirdropRecord is the record stream output and contains both ID and value.
`tokenAirdrop` isn't changed, except that it outputs `PendingAirdropRecord`s, as you noted.
The Id is a key, so it's used in `claim` and `cancel`.
The PendingAirdrop is the value in state, and is somewhat degenerate because an NFT serial number is part of the ID (and needs to be for claim or cancel to work correctly).

PendingAirdropId contains exactly what's necessary to claim or cancel an airdrop, as well as minimizing data duplication, particularly in the block stream (which is apparently a critical concern).

The structure proposed minimizes the number of times things get updated as well as avoiding possibly unlimited lists (we did discuss combining serial numbers into a list for the pending entry), which reduces block stream repeats (the entire item is in the block stream again when it's updated; rather different from record stream, which seems to often elide the state changes and expect mirror node to recreate them from transactions).

I'm preparing a change to this PR that excludes the changes to the protobuf structure (which is still necessary in some form prior to HIP approval, but can be examined separately from the other changes).  I'll then follow this PR with one just for the several protobuf changes for further discussion.